### PR TITLE
Tests for loader class and helper functions

### PR DESCRIPTION
**Pull request by mint-thompson**
**Description:**
This PR adds tests for the BasePackageLoader class. Tests are also added for the helper functions defined in `DefaultPackageLoader.ts`.

Additionally, due to my foolhardy clumsiness with git, I merged `main` into this branch, so this PR also includes the various documentation changes made during the transition to HL7.

**Testing Instructions:**
Run the tests locally and ensure they pass. Check the coverage report and verify that there is sufficient coverage in `BasePackageLoader.ts` and `DefaultPackageLoader.ts`. If there are test scenarios that should be added in order to cover more use cases, please let me know.

**Related Issue:**
N/A

### CODE REVIEW

File: src/loader/BasePackageLoader.ts
Comment (cmoesel) [inline src/loader/BasePackageLoader.ts]: Since this is an expected scenario (some packages contain non-resource JSON files), I think we should log this at the `debug` level rather than `info`.

File: test/loader/fixtures/StructureDefinition-Address.json
Comment (cmoesel) [inline test/loader/fixtures/StructureDefinition-Address.json:1]: It looks like most (or all?) of our StructureDefinitions use formatted JSON for readability. I'd suggest doing the same with this one.

File: test/loader/fixtures/StructureDefinition-human-being-logical-model.json
Comment (cmoesel) [inline test/loader/fixtures/StructureDefinition-human-being-logical-model.json]: Usually, if there's a top-level `text` property with a long `div`, we usually delete that `text` property -- as it's not important for our use cases. So you can delete `text` (the whole text object, including its `status` and `div` children).

File: test/loader/fixtures/StructureDefinition-valued-observation.json
Comment (cmoesel) [inline test/loader/fixtures/StructureDefinition-valued-observation.json]: I'd delete this gigantic `text` property too.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: When we load a `current` package, the stats come back with version `current` -- so that's probably what should happen in this mock, which also affects the log message assertion on line 119 below.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: This test is described as:
> should successfully download current version to cache that is missing or stale

but the expected message explicitly says the cached package date _matches_ the last build date -- which means the cached package is _not_ stale, so it should not attempt to download the build.

It looks like the test on line 192 already tests _not_ downloading a current build when it's not stale -- so I think you want to update this test to use a stale package date so it's detected as stale and downloads the current build.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: I don't understand this test. Support for specific branches is a feature of `current` builds (e.g., `some.ig#current$specific-branch`) -- but this is not testing current builds, it's testing registries. So I don't think branches are relevant. I think this is basically just a normal registry test with a weird version string. So... I don't think it's needed. It can probably be removed.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: Again, branches are only relevant for current builds... they do not apply to registries. So you can remove "without branch" from the description of this test.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:271]: This is an interesting one. A patch version like `1.2.x` will never be in the cache since we store it in the cache w/ its real version number. So it looks like we redownload it every time. Ideally, we probably ought to resolve `1.2.x` to a real version number and then check the cache. But... your test reflects the reality of how this is really done in the current code, so... there's not change to make in this test right now. It's testing the right thing.

(In the future though, we may want to improve the actual FPL behavior to be more efficient and avoid unnecessary downloads).

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:289]: Same comments apply for `latest` as for patch versions above. Again, no changes needed at this time -- but probably not the optimal implementation (oops).

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [reply test/loader/BasePackageLoader.test.ts]: Actually, wait... is this already covered in the test on line 365 below?
> should successfully download package when current version in cache out of date

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:191]: Is this basically testing the same use case as in line 409 below?
> should successfully use cached package when current version in cache is up to date

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:414]: Should we also check that `currentBuildClientMock.downloadCurrentBuild` _was_ called? Since that's how we know the cached package was treated as stale?

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:441]: Should we also check that `currentBuildClientMock.downloadCurrentBuild` _was_ called? Since that's how we know the cached package was treated as stale?

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: Is the package path supposed to be the same as the `packageJSONPath`? I'm assuming no.

Comment (cmoesel) [reply test/loader/BasePackageLoader.test.ts]: I see some others like this too...

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:532]: It seems like there are a series of tests here whose basic intent is to test the code in the private `loadResource` method to ensure that all the ResourceInfo is extracted and loaded correctly. That being the case, it seems there is probably a lot of duplicate code across these tests. Is there an opportunity to create a common test function that basically takes in a resource path and the expected ResourceInfo that should be produced upon loading?

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts:761]: As noted in the production code for this, I actually think this should be a `debug` message instead since it's an expected (and fairly common) situation.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: Delete this commented out code.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: Delete this code if it is not needed.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: Delete this code if it is not needed.

File: test/loader/BasePackageLoader.test.ts
Comment (cmoesel) [inline test/loader/BasePackageLoader.test.ts]: I don't think this test description is right.

**Review by cmoesel — CHANGES_REQUESTED** [review]
Wow. Those are some pretty intense tests! And dang, that's a lot of mocking! Nice job on these. I had a few questions and suggestions, but these are in pretty good shape!

File: test/loader/BasePackageLoader.test.ts
Comment (mint-thompson) [reply test/loader/BasePackageLoader.test.ts]: It looks like the test on line 192 tests for when the cache is already up-to-date, and the test on line 365 tests for when the cache is not up-to-date, so I'll remove the test on line 123.

File: test/loader/BasePackageLoader.test.ts
Comment (mint-thompson) [reply test/loader/BasePackageLoader.test.ts:191]: It is! But they seem to have slightly different assertions, so I'll combine those together.

File: test/loader/BasePackageLoader.test.ts
Comment (mint-thompson) [reply test/loader/BasePackageLoader.test.ts]: That's... very true. I'll remove these two tests.

File: test/loader/BasePackageLoader.test.ts
Comment (mint-thompson) [reply test/loader/BasePackageLoader.test.ts]: You are correct: Those should be using `packageBasePath`. I'll clean those up.

**Review by cmoesel — APPROVED** [review]
Thank you, @mint-thompson and @KaelynJefferson for this excellent set of tests!

Let's get a review from @jafeltra -- but please don't merge yet because we will want to do one more interactive rebase to tidy up the commits before doing a rebase-style merge into new-fpl.

**Pull request merged**
